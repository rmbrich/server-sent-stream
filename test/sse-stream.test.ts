import {describe, expect, test} from '@jest/globals';

import EventSourceStream from '../src/index';

const makeTestStream = (inputText: string) => {
    const chunks = inputText.split('|');
    let i = 0;
    return new ReadableStream<Uint8Array>({
        pull(controller) {
            if (i >= chunks.length) {
                controller.close();
                return;
            }
            controller.enqueue(new TextEncoder().encode(chunks[i]));
            i++;
        }
    });
};

const readTestStream = <T>(stream: ReadableStream<T>) => {
    const reader = stream.getReader();

    const values: T[] = [];

    return new Promise<T[]>(resolve => {
        const push = () => {
            void reader.read().then(({done, value}) => {

                if (done) {
                    resolve(values);
                } else {
                    values.push(value);
                    push();
                }
            });
        };

        push();
    });
};

const testEndToEnd = (inputText: string) => {
    const testStream = makeTestStream(inputText);
    const sseStream = new EventSourceStream();
    testStream.pipeThrough(sseStream);
    return readTestStream(sseStream.readable);
};

const testAllLineEndings = (inputText: string) => {
    test('cr', async() => {
        const crText = inputText.replaceAll('\n', '\r');
        const result = await testEndToEnd(crText);
        expect(result).toMatchSnapshot();
    });

    test('lf', async() => {
        const result = await testEndToEnd(inputText);
        expect(result).toMatchSnapshot();
    });

    test('crlf', async() => {
        const crlfText = inputText.replaceAll('\n', '\r\n');
        const result = await testEndToEnd(crlfText);
        expect(result).toMatchSnapshot();
    });
};

describe('makeSSEStream', () => {
    describe('basic', () => {
        testAllLineEndings('data: abc\ndata: def\n\ndata:foo\n\n');
    });

    describe('chunked', () => {
        testAllLineEndings('data: abc\ndata: def|\n\ndata:foo\n\n');
    });

    describe('very chunked', () => {
        testAllLineEndings('da|ta:| ab|c\ndata: |def|\n|\n|data:foo\n\n');
    });

    test('chunk split at crlf', async() => {
        const result = await testEndToEnd('data: abc\r|\ndata: def\r\n\r|\ndata:foo\r\n\r\n');
        expect(result).toMatchSnapshot();
    });

    describe('chunk split at end', () => {
        testAllLineEndings('data: abc\ndata: def|\n\ndata:foo\n|\n|');
    });

    describe('chunk split at space', () => {
        testAllLineEndings('data: |abc\ndata:| def|\n\ndata:|foo\n\n');
    });

    describe('only one leading space stripped', () => {
        testAllLineEndings('data:  abc\n\n');
    });

    describe('invalid stream', () => {
        testAllLineEndings('this will never match');
    });

    describe('only one trailing newline stripped', () => {
        testAllLineEndings('data:  abc\ndata:\ndata:\n\n');
    });

    describe('field with no colon', () => {
        testAllLineEndings('data: abc\ndata\ndata\ndata:foo\ndata\n\n');
    });

    describe('event types', () => {
        testAllLineEndings('data: foo\nevent: bar\nevent: baz\ndata: quux\n\ndata: noeventtype\n\n');
    });

    describe('event IDs', () => {
        // eslint-disable-next-line max-len
        testAllLineEndings('data: foo\nid: bar\nid: baz\ndata: quux\n\ndata: sameid\n\nid: newid\n\ndata: replacedid\n\n');
    });
});
