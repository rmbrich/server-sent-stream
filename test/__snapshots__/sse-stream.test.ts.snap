// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeSSEStream basic cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream basic crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream basic lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at end cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at end crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at end lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at space cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at space crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunk split at space lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunked cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunked crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream chunked lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream event IDs cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo
quux",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "baz",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "sameid",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "baz",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "replacedid",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "newid",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream event IDs crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo
quux",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "baz",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "sameid",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "baz",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "replacedid",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "newid",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream event IDs lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo
quux",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "baz",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "sameid",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "baz",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "replacedid",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "newid",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream event types cr 1`] = `
[
  MessageEvent {
    Symbol(type): "baz",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo
quux",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "noeventtype",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream event types crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "baz",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo
quux",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "noeventtype",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream event types lf 1`] = `
[
  MessageEvent {
    Symbol(type): "baz",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo
quux",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "noeventtype",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream field with no colon cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc


foo
",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream field with no colon crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc


foo
",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream field with no colon lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc


foo
",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream invalid stream cr 1`] = `[]`;

exports[`makeSSEStream invalid stream crlf 1`] = `[]`;

exports[`makeSSEStream invalid stream lf 1`] = `[]`;

exports[`makeSSEStream only one leading space stripped cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): " abc",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream only one leading space stripped crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): " abc",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream only one leading space stripped lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): " abc",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream only one trailing newline stripped cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): " abc

",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream only one trailing newline stripped crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): " abc

",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream only one trailing newline stripped lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): " abc

",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream very chunked cr 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream very chunked crlf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;

exports[`makeSSEStream very chunked lf 1`] = `
[
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "abc
def",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
  MessageEvent {
    Symbol(type): "message",
    Symbol(kTarget): null,
    Symbol(kIsBeingDispatched): false,
    Symbol(kData): "foo",
    Symbol(kOrigin): "",
    Symbol(kLastEventId): "",
    Symbol(kSource): null,
    Symbol(kPorts): [],
  },
]
`;
